apis/s3/src/main/java/org/jclouds/s3/xml/ListAllMyBucketsHandler.java

```diff
@@ -56,9 +57,18 @@ public ListAllMyBucketsHandler(DateService dateParser) {
 
    public void endElement(String uri, String name, String qName) {
       if (qName.equals("ID")) { // owner stuff
-         currentOwner = new CanonicalUser(currentOrNull(currentText));
+         if (currentDisplayName == null) {
+            currentOwner = new CanonicalUser(currentOrNull(currentText));
+         } else {
+            currentOwner = new CanonicalUser(
+               currentOrNull(currentText), currentDisplayName);
+         }
       } else if (qName.equals("DisplayName")) {
-         currentOwner.setDisplayName(currentOrNull(currentText));
+         if (currentOwner == null) {
+            currentDisplayName = currentOrNull(currentText);
+         } else {
+            currentOwner.setDisplayName(currentOrNull(currentText));
```

Comment:
The original code is wrong here.  Instead what I think would improve it is storing `id` and `displayName` as `String`s then putting this conditional in the Bucket clause.  Note that you will want to set these fields to `null` as well.

Reply:
I see, yes, that's better, although in the current patch and suggested patch there is the latent bug of having the `<Owner>` element appear _after_ the bucket listing. What do you think about adding an empty constructor for the `CanonicalUser` class and setting the values once they are parsed?

Reply:
I updated the patch with a test for a reverse order of <Buckets> and <Owner> tags and expanded the CanonicalUser class to allow setting the ID and a public constructor. This allowed me to change the parser to avoid relying on the order of the elements when processing the XML.